enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+
+### Open Questions
+
+I've put down the options as I see them. I need some input from the community to figure out the right way to go.
+The first two would required breaking changes and therefore migration to a new schema. The [first option](#Union-types)
+is better assuming we don't want to give users the option to implement controllers for additional, and custom,
+`CatalogSource` types. Otherwise, the [seconds option](#CatalogSourceClass) would be better. If we have bigger changes on
+the horizon, we may want to punt on the "right" approach, in favour of the [third option](#ImageConfig). This would 
+mitigate the issue, but not necessarily in the best way.
+
+I'm personally in favor of options 1 or 2, even though they will require more work, I think they provide an improvement
+in user experience.
+
+## Summary
+
+Currently, there is no way to override the pod spec for `grpc` type `CatalogSource` pods.
+This is desirable, for example, in cases where the admin would like additional control of 
+where in the infrastructure the catalog source pod get deployed. In this PE we
+propose changes to the `CatalogSource` spec to enable this kind customization.
+
+
+## Motivation
+
+This EP targets issue [#2452](https://github.com/operator-framework/operator-lifecycle-manager/issues/2452).
+Users would like to be able to override the pod spec of the CatalogSource pod
+created for `grpc` `CatalogSource`s. Especifically, they would like to be able to specify
+pod tolerations and node selectors.
+
+### Goals
+
+ * Update the `CatalogSource` spec to allow users to affect the pod spec of `grpc` `CatalogSource` pods 
+
+### Non-Goals
+
+ TBD
+
+## Proposal
+
+### Current State
+
+The [current](https://github.com/operator-framework/api/blob/a80624eab36b2005fa5d84c9b539bb4817e44daa/crds/operators.coreos.com_catalogsources.yaml#L54) `CatalogSource.spec` accepts three difference values for the `spec.sourceType`: `grpc`, `configmap`, and `internal`. Depending on the value, different optional attributes are used to deploy the underlying infrastructure to serve the catalog content. For instance, if `spec.sourceType` is set to `grpc`, the user can set either the `spec.address` or `spec.image`. If both are set, `spec.image` takes precedence. If `spec.sourceType` is set to `configmap`,
+`spec.configMap` gets used, but `spec.address` and `spec.image` get ignored. Since there could be attributes defined that just get ignored by the controller, it could be hard for a user to know what is happening given just a `CatalogSource` CR and no documentation.
+
+### Options
+
+Three different options are outlined below. The first two would require breaking changes to the API, while the third would be the "minimal" optional. The path taken likely depends on the whether we are planning for a new `CatalogSource` CRD version in the near future or not.
```

Comment:
We probably need to define here a plan if we decide to move forward with a breaking change, so in this case we need to:
- What/who will be impacted?
- How to be adapt to the new changes? 
- What are the APIs/endpoints which must be deprecated? Can we warn for a grace period?

Reply:
Oh, absolutely. This EP is faaar from finished. What I'm looking for _now_ though is some signal from the community to tell me that doing it the "right" way (with breaking changes) is worth the investment at the moment. Or, if we can get away with something quick and dirty to unblock a certain section of users, and punt on breaking changes now since we can see there will be more coming in the near future (and then do everything at once). Do you know what I mean?

Reply:
I may be missing something but my understanding is that you would only add optional fields. If users don't leverage them the behaviour stays the same. They are also fields that would be set by end users so that it would be acceptable to have them available from a specific OLM version. It is different when things get provided by operator authors that may not know the target OLM version beforehands.
I believe that it should be what is described in Kubernetes as [a "compatible" change](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-a-field). There are also a few [recommendations](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) for adding "unstable" features to a stable version. 
My understanding is that following the recommendations the impact on users not leveraging the feature should be zero. Users that decide to leverage it should be made aware that it is not stable. Am I missing your point?

Reply:
I don't know if I'm answering your concern. But, it depends on the solution path we take. For the minimal solution, it would just be an additional optional field and the CRD version can stay the same. If we go for a more comprehensive option, we will have to deprecate some fields, then eventually, once these fields are removed, bump the version. In the meantime, it would be as you describe as well. Additional optional fields that get activated in certain circumstances.